litestar/contrib/msgspec.py

```diff
@@ -0,0 +1,60 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Generic, TypeVar, cast
+
+from msgspec import NODEFAULT, Struct, inspect
+
+from litestar.dto.factory.abc import AbstractDTOFactory
+from litestar.dto.factory.field import DTO_FIELD_META_KEY, DTOField
+from litestar.dto.factory.types import FieldDefinition
+from litestar.dto.factory.utils import get_model_type_hints
+from litestar.types.empty import Empty
+from litestar.utils.helpers import get_fully_qualified_class_name
+
+if TYPE_CHECKING:
+    from typing import Any, ClassVar, Collection, Generator
+
+    from litestar.typing import ParsedType
+
+__all__ = ("MsgspecDTO",)
+
+DataT = TypeVar("DataT", bound="Struct | Collection[Struct]")
+
+
+class MsgspecDTO(AbstractDTOFactory[DataT], Generic[DataT]):
+    """Support for domain modelling with Msgspec."""
+
+    __slots__ = ()
+
+    model_type: ClassVar[type[Struct]]
+
+    @classmethod
+    def generate_field_definitions(cls, model_type: type[Struct]) -> Generator[FieldDefinition, None, None]:
+        msgspec_fields = {f.name: f for f in cast("inspect.StructType", inspect.type_info(model_type)).fields}
+
+        def default_or_empty(value: Any) -> Any:
+            if value is NODEFAULT:
+                return Empty
+            return value
+
+        for key, parsed_type in get_model_type_hints(model_type).items():
+            msgspec_field = msgspec_fields[key]
+
+            dto_field: DTOField | None = None
+            if isinstance(msgspec_field.type, inspect.Metadata):
+                dto_field = (msgspec_field.type.extra or {}).get(DTO_FIELD_META_KEY)
+
+            field_def = FieldDefinition(
```

Comment:
So one thing to consider is to access and copy any field specific constraints defined in msgspec as `Annotated[str, msgspec.Meta(...)]`. There is logic in place added to `SignatureField.from_type`, but we will need the same for DTOs - it would also need to affect documentation. 

One possible solution is to copy msgspec.Meta using `ParameterKwarg` or `BodyKwarg`. Or create another layer of representation, or copy the type over and worry about this in Signature field.

Reply:
Yes, I want to do this with all of the source model types so I'll do in another PR as part of the 2.0 milestone work.